jetty-server/src/main/java/org/eclipse/jetty/server/MultiPartFormInputStream.java

```diff
@@ -53,24 +52,49 @@
  * MultiPartInputStream
  * <p>
  * Handle a MultiPart Mime input stream, breaking it up on the boundary into files and strings.
- *
+ * </p>
+ * <p>
+ * Deleting the parts can be done from a different thread if the parts are parsed asynchronously.
+ * Because of this we use the state to fail the parsing and coordinate which thread will delete any remaining parts.
+ * The deletion of parts is done by the cleanup thread in all cases except the transition from ERROR-&gt;DELETED which
+ * is done by the parsing thread.
+ * </p>
+ * <pre>{@code
+ *                              deleteParts()
+ *     +--------------------------------------------------------------+
+ *     |                                                              |
+ *     |                                          deleteParts()       v
+ *  UNPARSED -------> PARSING --------> PARSED  ------------------>DELETED
+ *                      |                                             ^
+ *                      |                                             |
+ *                      +----------------> ERROR ---------------------+
+ *                        deleteParts()             parsing thread
+ * }</pre>
  * @see <a href="https://tools.ietf.org/html/rfc7578">https://tools.ietf.org/html/rfc7578</a>
  */
 public class MultiPartFormInputStream
 {
+    private enum State
+    {
+        UNPARSED,
+        PARSING,
+        PARSED,
+        DELETED,
+        ERROR
+    }
+
     private static final Logger LOG = Log.getLogger(MultiPartFormInputStream.class);
-    private static final MultiMap<Part> EMPTY_MAP = new MultiMap<>(Collections.emptyMap());
-    private final MultiMap<Part> _parts;
-    private InputStream _in;
-    private MultipartConfigElement _config;
-    private String _contentType;
-    private Throwable _err;
-    private File _tmpDir;
-    private File _contextTmpDir;
-    private boolean _deleteOnExit;
-    private boolean _writeFilesWithFilenames;
-    private boolean _parsed;
-    private int _bufferSize = 16 * 1024;
+    private final MultiMap<Part> _parts = new MultiMap<>();
+    private final InputStream _in;
+    private final MultipartConfigElement _config;
+    private final File _contextTmpDir;
+    private final String _contentType;
+    private volatile Throwable _err;
```

Comment:
Are you sure all these need to be volatiles? Also the state variable - aren't you always reading/writing that in a sync block?

Reply:
I'm not sure if they need to be volatile I just marked them all to be safe. 
The state is assigned outside synchronized in the constructor so wasn't sure if I should declare volatile.